From cf906bb2f6b1b58c8e39ee39c7d00258fe491fcc Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Date: Wed, 31 Aug 2022 11:18:21 +0300
Subject: [PATCH 087/102] misc: s32cc_cmu: Include FIRC drift in 'Expected
 range'

Include +-5% frequency drift of the FIRC clock according to
S32G and S32R reference manuals.

Issue: ALB-9151
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/misc/s32cc_cmu.c | 11 +++++++----
 drivers/misc/s32cc_cmu.h |  8 +++++++-
 2 files changed, 14 insertions(+), 5 deletions(-)

diff --git a/drivers/misc/s32cc_cmu.c b/drivers/misc/s32cc_cmu.c
index 71ca2be1d5..ee2687dc45 100644
--- a/drivers/misc/s32cc_cmu.c
+++ b/drivers/misc/s32cc_cmu.c
@@ -4,8 +4,8 @@
  */
 
 #include <common.h>
-#include <command.h>
 #include <clk.h>
+#include <command.h>
 #include <dm.h>
 #include <inttypes.h>
 #include <misc.h>
@@ -90,7 +90,8 @@ struct cmu_result {
 static struct cmu s32g2_cmu_blocks[] = {
 	FIRC_PERIPH_CMU_FC(0, FXOSC, FXOSC_FREQ),
 
-	FXOSC_PERIPH_CMU_FM(1, FIRC, FIRC_FREQ),
+	/* +-5% drift due to silicon and temperature factors  */
+	FXOSC_PERIPH_CMU_FM_RANGE(1, FIRC, FIRC_MIN_FREQ, FIRC_MAX_FREQ),
 	FXOSC_PERIPH_CMU_FM(2, SIRC, SIRC_FREQ),
 	FXOSC_PERIPH_CMU_FM(3, FTM_0_REF, 40000000),
 	FXOSC_PERIPH_CMU_FM(4, FTM_1_REF, 40000000),
@@ -134,7 +135,8 @@ static struct cmu s32g2_cmu_blocks[] = {
 static struct cmu s32g3_cmu_blocks[] = {
 	FIRC_PERIPH_CMU_FC(0, FXOSC, FXOSC_FREQ),
 
-	FXOSC_PERIPH_CMU_FM(1, FIRC, FIRC_FREQ),
+	/* +-5% drift due to silicon and temperature factors  */
+	FXOSC_PERIPH_CMU_FM_RANGE(1, FIRC, FIRC_MIN_FREQ, FIRC_MAX_FREQ),
 	FXOSC_PERIPH_CMU_FM(2, SIRC, SIRC_FREQ),
 	FXOSC_PERIPH_CMU_FM(3, FTM_0_REF, 40000000),
 	FXOSC_PERIPH_CMU_FM(4, FTM_1_REF, 40000000),
@@ -178,7 +180,8 @@ static struct cmu s32g3_cmu_blocks[] = {
 static struct cmu s32r45_cmu_blocks[] = {
 	FIRC_PERIPH_CMU_FC(0, FXOSC, FXOSC_FREQ),
 
-	FXOSC_PERIPH_CMU_FM(1, FIRC, FIRC_FREQ),
+	/* +-5% drift due to silicon and temperature factors  */
+	FXOSC_PERIPH_CMU_FM_RANGE(1, FIRC, FIRC_MIN_FREQ, FIRC_MAX_FREQ),
 	FXOSC_PERIPH_CMU_FM(2, SIRC, SIRC_FREQ),
 	FXOSC_PERIPH_CMU_FM(3, FTM_0_REF, 40000000),
 	FXOSC_PERIPH_CMU_FM(4, FTM_1_REF, 40000000),
diff --git a/drivers/misc/s32cc_cmu.h b/drivers/misc/s32cc_cmu.h
index 6cd7825139..51452fa823 100644
--- a/drivers/misc/s32cc_cmu.h
+++ b/drivers/misc/s32cc_cmu.h
@@ -8,11 +8,13 @@
 
 #include <linux/stringify.h>
 
-#define FIRC_VARIATION			(60ULL)	// 6.0%
+#define FIRC_VARIATION			(50ULL)	// 5.0%
 #define FXOSC_VARIATION			(5ULL)	// 0.5%
 #define PERIPH_VARIATION		(5ULL)	// 0.5%
 
 #define FIRC_FREQ			(48000000ULL)
+#define FIRC_MIN_FREQ			(45600000ULL)
+#define FIRC_MAX_FREQ			(50400000ULL)
 #define FXOSC_FREQ			(40000000ULL)
 #define SIRC_FREQ			(32000ULL)
 
@@ -63,6 +65,10 @@
 	CMU(ID, FXOSC, MON, FXOSC_FREQ, MON_FRQ, \
 			0, 0, false)
 
+#define FXOSC_PERIPH_CMU_FM_RANGE(ID, MON, MIN_FRQ, MAX_FRQ) \
+	CMU_EXP_RANGE(ID, FXOSC, MON, FXOSC_FREQ, MIN_FRQ, MAX_FRQ, \
+			FXOSC_VARIATION, PERIPH_VARIATION, false)
+
 #define FIRC_PERIPH_CMU_FM(ID, MON, MON_FRQ) \
 	CMU(ID, FIRC, MON, FIRC_FREQ, MON_FRQ, \
 			0, 0, false)
-- 
2.17.1

